[Generic signature builder] Synchronize representative with the archetype anchor

DougGregor opened the pull request:
Make the representative into the archetype anchor. This eliminates an annoying walk over the equivalence class whenever we want to extract the archetype anchor, as well as eliminating a conceptual hindrance in the generic signature builder.

DougGregor committed:
[Generic signature builder] Ordering of nested types doesn't matter.

... so don't treat typealias-bound nested types as special.

DougGregor committed:
[Generic signature builder] Align isBetterArchetypeAnchor and compareDependentTypes.

The latter was also performing the concrete-type-in-path check; do
that within compareDependentTypes(); we don't want a different here.

DougGregor committed:
[Generic signature builder] Align the representative with the archetype anchor.

Make the representative of an equivalence class also be the archetype
anchor, so we don't need to walk through the equivalence class each
time we need the archetype anchor.

DougGregor committed:
[Generic signature builder] Eliminate isBetterArchetypeAnchor.

It's just a wrapper around compareDependentTypes() that was only used
for assertions.

DougGregor commented:
Note that this breaks given:

  class A<T, U> where T : Sequence, U : Sequence, T.Iterator.Element == U.Iterator.Element {
    func foo<V>(v: V) where T == Array<Int> {
    }
  }

slavapestov commented:
@DougGregor Wasn't that already broken? It drops the same-type requirement when building the innermost signature.

DougGregor commented:
It breaks... harder? I'm going to fix that issue first and then re-try this PR.

huonw commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -1349,6 +1317,12 @@ bool GenericSignatureBuilder::addSameTypeRequirementBetweenArchetypes(
     T1->ConcreteTypeSource = T2->ConcreteTypeSource;
   }
 
+  // Decide which potential archetype is to be considered the representative.
```
I'm curious why this needs to move, the code above seems like it shouldn't care?

DougGregor replied:
`compareDependentTypes` checks for a concrete type in the path of the potential archetype (which includes the current PA), so we need either both or neither to have a concrete type so that we effectively skip that check.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Generic signature builder] Concrete constraints across equivalence classes.

When emitting a requirement for a same-type-to-concrete constraint,
use the known concrete source for just the first of the component
anchors. For the rest, we need to explicitly model the constraint lest
it get lost. Fixes rdar://problem/30478915.

DougGregor committed:
[Generic signature builder] Eliminate "concrete type in path" checks.

We no longer need the "concrete type in path" checks to correctly
order potential archetypes, and they prohibit us from making the
representative always the archetype anchor.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge
